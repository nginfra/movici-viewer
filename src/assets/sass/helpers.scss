.is-size-6-half {
  font-size: 0.85rem !important;
}

.text-ellipsis {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: pre;
}

// .x-overflow is used alongside with FixedPosition.ts as the overflow parent
//
.overflow,
.x-overflow {
  overflow: auto;
  padding-right: 0.5rem;
  margin-right: -0.5rem;
}

.overflow-hidden {
  overflow: hidden !important;
}

.overflow-hover {
  overflow: hidden;
  &:hover {
    overflow: auto;
    padding-right: 0.5rem;
    margin-right: -0.5rem;
  }
}

button {
  &.is-borderless {
    border: 0;
  }
  &.is-transparent {
    background-color: transparent;
    color: $black;
  }
  &.is-round {
    border-radius: 100%;
  }
  &.movici {
    background: $white;
    color: $primary;
    border: 2px solid $white-ter !important;
    &:hover {
      color: $primary;
      background: $primary-light;
      border-color: $grey-light;
    }
    &:active {
      color: $primary;
      border-color: $primary !important;
    }
  }
}

[required] {
  label,
  .label {
    &::after {
      margin-left: 2px;
      content: '*';
      color: $grey;
      font-weight: bold;
    }
  }
}
