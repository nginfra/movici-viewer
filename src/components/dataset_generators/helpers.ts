// @ts-nocheck
import GeoJSON from 'ol/format/GeoJSON';
import { transformGeoJsonToCRS } from '@/crs';
import { MoviciError } from '@/errors';
import { DatasetGeneratorDataset } from '@/types';

export function findDataset(type: string, currentDataset: DatasetGeneratorDataset) {
  let currentChild, result;

  if (type === currentDataset.type) {
    return currentDataset;
  } else {
    if (currentDataset.childs === undefined) {
      return null;
    }

    // Loop over child types
    for (let childTypeIdx = 0; childTypeIdx < currentDataset.childs.length; ++childTypeIdx) {
      const currentChildType = currentDataset.childs[childTypeIdx];

      for (let childIdx = 0; childIdx < currentChildType.datasets.length; ++childIdx) {
        currentChild = currentChildType.datasets[childIdx];
        // Search in the current child
        result = findDataset(type, currentChild);

        if (result !== null) {
          return result;
        }
      }

      // Return the result if the node has been found
    }

    // The node has not been found and we have no more options
    return null;
  }
}

export function validateGeoJSON(textData: string) {
  let feature;
  try {
    feature = new GeoJSON().readFeature(textData);
  } catch (e) {
    console.warn('Invalid GeoJSON supplied');
    failMessage('Invalid GeoJSON supplied');
    return false;
  }

  const featureType = feature?.getGeometry()?.getType();
  if (featureType?.toLowerCase() == 'polygon') {
    return true;
  }

  failMessage(`Feature is not a Polygon but a ${featureType}`);
  return false;
}

export function transformGeoJSON(geojson: GeoJSON) {
  let transformed;
  try {
    transformed = transformGeoJsonToCRS(geojson);
  } catch (e) {
    console.warn(e.message);
    if (e instanceof MoviciError) {
      failMessage(e.message);
    }
  }
  return transformed;
}
