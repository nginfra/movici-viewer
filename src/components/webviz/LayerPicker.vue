<template>
  <div class="layer-picker-container is-flex is-flex-direction-column" :class="{ reduce }">
    <EditLayerModal
      :active="showLayerDetails"
      @close="showLayerDetails = false"
      :value="currentItem"
      @input="updateCurrentItem($event)"
      @delete="deleteCurrentItem"
      :mode="mode"
    />
    <div class="overflow">
      <LayerPickerCard
        :class="{ 'mb-4': idx < value.length - 1 }"
        v-for="(layer, idx) in value"
        :mode="mode"
        :key="idx"
        :reduce="reduce"
        :value="layer"
        @input="updateItem(idx, $event)"
        @delete="deleteItem(idx)"
        @edit="startEditingItem(idx)"
      ></LayerPickerCard>
    </div>
    <div class="buttons mt-4 is-flex-shrink" :class="reduce ? 'is-centered' : 'is-right'">
      <BButton class="is-primary" @click="startEditingItem(null)" icon-left="plus">
        <template v-if="!reduce">Add data layer</template>
      </BButton>
    </div>
  </div>
</template>

<script lang="ts">
import LayerPickerCard from './LayerPickerCard.vue';
import EditLayerModal from './EditLayerModal.vue';
import { Component, Prop, Vue } from 'vue-property-decorator';
import { VisualizationMode } from '@/types';
import { VisualizerInfo } from '@/visualizers';

@Component({
  name: 'LayerPicker',
  components: { LayerPickerCard, EditLayerModal }
})
export default class LayerPicker extends Vue {
  @Prop({
    type: Boolean,
    default: false
  })
  readonly reduce!: boolean;

  @Prop({
    type: String,
    default: VisualizationMode.GEOMETRY,
    validator(value) {
      return Object.values(VisualizationMode).includes(value);
    }
  })
  readonly mode!: VisualizationMode;

  @Prop({
    type: Array,
    default: () => []
  })
  readonly value!: VisualizerInfo[];

  showLayerDetails = false;

  currentIndex: number | null = null;

  get currentItem(): VisualizerInfo | null {
    return this.currentIndex === null ? null : this.value[this.currentIndex];
  }

  startEditingItem(index: number | null) {
    this.currentIndex = index;
    this.showLayerDetails = true;
  }

  updateCurrentItem(val: VisualizerInfo | null) {
    if (val) {
      if (this.currentIndex === null) {
        this.currentIndex = this.value.length;

        this.$emit('input', [...this.value, val]);
      } else {
        this.updateItem(this.currentIndex, val);
      }
    }
  }

  updateItem(idx: number, val: VisualizerInfo) {
    this.$emit(
      'input',
      this.value.map((info, arrayIdx) => {
        return arrayIdx === idx ? val : info;
      })
    );
  }

  deleteCurrentItem() {
    this.currentIndex !== null && this.deleteItem(this.currentIndex);
  }

  deleteItem(idx: number) {
    this.$emit(
      'input',
      this.value.filter((val, arrayIdx) => idx !== arrayIdx)
    );
  }
}
</script>

<style scoped lang="scss">
.buttons {
  background-color: $white-ter;
}
.layer-picker-container {
  height: calc(100vh - 348px);
  &.reduce {
    height: calc(100vh - 264px);
  }
}
</style>
