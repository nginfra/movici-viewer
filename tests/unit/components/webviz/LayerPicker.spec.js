import LayerPicker from '@/components/webviz/LayerPicker';
import { VisualizerInfo } from '@/visualizers';
import { createComponentWrapper } from '../../helpers';

describe('LayerPicker.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = createComponentWrapper(LayerPicker, {
      propsData: {
        value: [
          new VisualizerInfo({
            name: 'my_dataset',
            entityGroup: 'electrical_node_entities',
            componentProperty: 'operational_status_properties/is_working_properly',
            visible: true
          })
        ]
      }
    });
  });
  it('renders correctly for geometryOnly layer', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('emits data on update', async () => {
    const sentinel = { key: 'sentinel' };
    await wrapper.findComponent({ name: 'LayerPickerCard' }).vm.$emit('input', sentinel);
    expect(wrapper.emitted().input[0]).toStrictEqual([[sentinel]]);
  });
});
