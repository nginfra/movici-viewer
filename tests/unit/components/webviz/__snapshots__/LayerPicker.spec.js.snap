// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerPicker.vue renders correctly for geometryOnly layer 1`] = `
<div class="layer-picker-container is-flex is-flex-direction-column">
  <transition-stub name="zoom-out">
    <!---->
  </transition-stub>
  <div class="overflow">
    <div class="">
      <div class="box has-text-centered" style="display: none;"><span class="icon"><i class="fas fa-project-diagram"></i></span></div>
      <div class="box is-flex-space-between">
        <div><span><strong>my_dataset</strong></span>
          <!---->
        </div>
        <div>
          <!----> <button class="button is-white is-small"><span class="icon has-text-primary is-medium"><i class="fas fa-eye fa-lg"></i></span></button> <span class="b-tooltip is-black is-top is-medium"><transition-stub name="fade"><div class="tooltip-content" style="display: none;">Edit</div></transition-stub><div class="tooltip-trigger"><button type="button" class="button is-borderless action-edit is-small"><!----><span><span class="icon has-text-info is-medium"><i class="fas fa-pen fa-lg"></i></span></span>
          <!----></button>
        </div></span><span class="b-tooltip is-black is-top is-medium"><transition-stub name="fade"><div class="tooltip-content" style="display: none;">Delete</div></transition-stub><div class="tooltip-trigger"><button type="button" class="button is-borderless action-delete is-small"><!----><span><span class="icon has-text-danger is-medium"><i class="fas fa-trash fa-lg"></i></span></span>
        <!----></button>
      </div></span>
    </div>
  </div>
</div>
</div>
<div class="buttons mt-4 is-flex-shrink is-right"><button type="button" class="button is-primary"><span class="icon is-small"><i class="fas fa-plus"></i></span><span>Add data layer</span>
    <!---->
  </button></div>
</div>
`;
